File: packages/database/src/schema/ai.ts
import {
  boolean,
  index,
  integer,
  jsonb,
  numeric,
  pgTable,
  text,
  timestamp,
  uniqueIndex,
  uuid,
  varchar,
} from 'drizzle-orm/pg-core';
import { aiSessionStatusEnum } from './enums';
import { extensions } from './telephony';
import { tenants } from './tenants';
import { users } from './users';

export const aiProviderConnections = pgTable(
  'ai_provider_connections',
  {
    id: uuid('id').primaryKey().defaultRandom(),
    tenantId: uuid('tenant_id')
      .notNull()
      .references(() => tenants.id, { onDelete: 'cascade' }),
    providerType: varchar('provider_type', { length: 64 }).notNull(),
    name: varchar('name', { length: 255 }).notNull(),
    credentialsEncrypted: text('credentials_encrypted').notNull(),
    credentialKeyVersion: varchar('credential_key_version', { length: 16 }).notNull().default('v1'),
    config: jsonb('config').notNull().default({}),
    isActive: boolean('is_active').notNull().default(true),
    healthStatus: varchar('health_status', { length: 32 }).notNull().default('unknown'),
    validationStatus: varchar('validation_status', { length: 32 }).notNull().default('unknown'),
    lastValidatedAt: timestamp('last_validated_at', { withTimezone: true }),
    validationError: text('validation_error'),
    createdBy: uuid('created_by').references(() => users.id),
    createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp('updated_at', { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [index('ai_provider_connections_tenant_idx').on(table.tenantId)],
);

export const aiAgents = pgTable(
  'ai_agents',
  {
    id: uuid('id').primaryKey().defaultRandom(),
    tenantId: uuid('tenant_id')
      .notNull()
      .references(() => tenants.id, { onDelete: 'cascade' }),
    name: varchar('name', { length: 255 }).notNull(),
    description: text('description'),
    status: varchar('status', { length: 32 }).notNull().default('draft'),
    routeNumber: varchar('route_number', { length: 16 }),
    transferExtensionId: uuid('transfer_extension_id').references(() => extensions.id),
    activeVersionId: uuid('active_version_id'),
    isActive: boolean('is_active').notNull().default(false),
    createdBy: uuid('created_by').references(() => users.id),
    createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp('updated_at', { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    index('ai_agents_tenant_idx').on(table.tenantId),
    uniqueIndex('ai_agents_tenant_route_uidx').on(table.tenantId, table.routeNumber),
  ],
);

export const aiAgentVersions = pgTable(
  'ai_agent_versions',
  {
    id: uuid('id').primaryKey().defaultRandom(),
    tenantId: uuid('tenant_id')
      .notNull()
      .references(() => tenants.id, { onDelete: 'cascade' }),
    agentId: uuid('agent_id')
      .notNull()
      .references(() => aiAgents.id, { onDelete: 'cascade' }),
    version: integer('version').notNull(),
    config: jsonb('config').notNull().default({}),
    pipelineType: varchar('pipeline_type', { length: 32 }).notNull().default('realtime'),
    providerConnectionId: uuid('provider_connection_id').references(() => aiProviderConnections.id),
    provider: varchar('provider', { length: 64 }),
    model: varchar('model', { length: 128 }),
    voice: varchar('voice', { length: 64 }),
    language: varchar('language', { length: 16 }).default('en'),
    systemInstructions: text('system_instructions'),
    openingMessage: text('opening_message'),
    interruptionConfig: jsonb('interruption_config').notNull().default({}),
    silenceTimeoutSeconds: integer('silence_timeout_seconds'),
    maxDurationSeconds: integer('max_duration_seconds'),
    allowedTools: jsonb('allowed_tools').notNull().default([]),
    recordingPolicy: varchar('recording_policy', { length: 32 }),
    transcriptionPolicy: varchar('transcription_policy', { length: 32 }),
    status: varchar('status', { length: 32 }).notNull().default('draft'),
    createdBy: uuid('created_by').references(() => users.id),
    createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
    activatedAt: timestamp('activated_at', { withTimezone: true }),
  },
  (table) => [
    uniqueIndex('ai_agent_versions_uidx').on(table.agentId, table.version),
    index('ai_agent_versions_tenant_idx').on(table.tenantId),
  ],
);

export const aiTools = pgTable(
  'ai_tools',
  {
    id: uuid('id').primaryKey().defaultRandom(),
    tenantId: uuid('tenant_id')
      .notNull()
      .references(() => tenants.id, { onDelete: 'cascade' }),
    name: varchar('name', { length: 128 }).notNull(),
    toolType: varchar('tool_type', { length: 64 }).notNull(),
    jsonSchema: jsonb('json_schema').notNull(),
    config: jsonb('config').notNull().default({}),
    requiresApproval: boolean('requires_approval').notNull().default(false),
    isActive: boolean('is_active').notNull().default(true),
    createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp('updated_at', { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    uniqueIndex('ai_tools_tenant_name_uidx').on(table.tenantId, table.name),
    index('ai_tools_tenant_idx').on(table.tenantId),
  ],
);

export const aiKnowledgeSources = pgTable(
  'ai_knowledge_sources',
  {
    id: uuid('id').primaryKey().defaultRandom(),
    tenantId: uuid('tenant_id')
      .notNull()
      .references(() => tenants.id, { onDelete: 'cascade' }),
    name: varchar('name', { length: 255 }).notNull(),
    sourceType: varchar('source_type', { length: 64 }).notNull(),
    config: jsonb('config').notNull().default({}),
    isActive: boolean('is_active').notNull().default(true),
    createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp('updated_at', { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [index('ai_knowledge_sources_tenant_idx').on(table.tenantId)],
);

export const aiSessions = pgTable(
  'ai_sessions',
  {
    id: uuid('id').primaryKey().defaultRandom(),
    tenantId: uuid('tenant_id')
      .notNull()
      .references(() => tenants.id, { onDelete: 'cascade' }),
    callId: uuid('call_id').notNull(),
    agentId: uuid('agent_id')
      .notNull()
      .references(() => aiAgents.id),
    agentVersionId: uuid('agent_version_id')
      .notNull()
      .references(() => aiAgentVersions.id),
    providerConnectionId: uuid('provider_connection_id').references(() => aiProviderConnections.id),
    providerType: varchar('provider_type', { length: 64 }).notNull(),
    providerSessionId: varchar('provider_session_id', { length: 256 }),
    status: aiSessionStatusEnum('status').notNull().default('connecting'),
    state: varchar('state', { length: 32 }).notNull().default('CREATED'),
    correlationId: uuid('correlation_id').notNull(),
    diagnostics: jsonb('diagnostics').notNull().default({}),
    transferResult: jsonb('transfer_result').notNull().default({}),
    failureCategory: varchar('failure_category', { length: 64 }),
    timing: jsonb('timing').notNull().default({}),
    startedAt: timestamp('started_at', { withTimezone: true }).notNull().defaultNow(),
    endedAt: timestamp('ended_at', { withTimezone: true }),
  },
  (table) => [
    index('ai_sessions_call_idx').on(table.callId),
    index('ai_sessions_tenant_idx').on(table.tenantId),
    index('ai_sessions_correlation_idx').on(table.correlationId),
  ],
);

export const aiUsage = pgTable(
  'ai_usage',
  {
    id: uuid('id').primaryKey().defaultRandom(),
    tenantId: uuid('tenant_id')
      .notNull()
      .references(() => tenants.id),
    sessionId: uuid('session_id')
      .notNull()
      .references(() => aiSessions.id),
    callId: uuid('call_id'),
    provider: varchar('provider', { length: 64 }).notNull(),
    meterName: varchar('meter_name', { length: 64 }).notNull(),
    quantity: numeric('quantity', { precision: 18, scale: 6 }).notNull(),
    unit: varchar('unit', { length: 32 }).notNull(),
    measurementSource: varchar('measurement_source', { length: 32 }).notNull().default('PLATFORM_MEASURED'),
    idempotencyKey: varchar('idempotency_key', { length: 128 }),
    providerEventId: varchar('provider_event_id', { length: 256 }),
    correlationId: uuid('correlation_id'),
    costAmount: numeric('cost_amount', { precision: 18, scale: 6 }),
    costCurrency: varchar('cost_currency', { length: 3 }).default('USD'),
    recordedAt: timestamp('recorded_at', { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    index('ai_usage_tenant_idx').on(table.tenantId),
    uniqueIndex('ai_usage_idempotency_uidx').on(table.idempotencyKey),
  ],
);

File: services/telephony-controller/internal/repository/ai.go
package repository

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
)

type AiAgentInfo struct {
	ID                   uuid.UUID
	TenantID             uuid.UUID
	TenantSlug           string
	RouteNumber          string
	TransferExtensionID  uuid.UUID
	TransferNumber       string
	ActiveVersionID      uuid.UUID
	ProviderConnectionID uuid.UUID
	Provider             string
	Model                string
	Voice                string
	Language             string
	SystemInstructions   string
	OpeningMessage       string
	AllowedTools         []string
	MaxDurationSeconds   int
}

type AiSessionRecord struct {
	ID                   uuid.UUID
	TenantID             uuid.UUID
	CallID               uuid.UUID
	AgentID              uuid.UUID
	AgentVersionID       uuid.UUID
	ProviderConnectionID uuid.UUID
	ProviderType         string
	CorrelationID        uuid.UUID
}

func (r *Repository) LookupAiAgentByRoute(ctx context.Context, tenantSlug, routeNumber string) (*AiAgentInfo, error) {
	var info AiAgentInfo
	var toolsJSON []byte
	err := r.withBypass(ctx, func(tx pgx.Tx) error {
		return tx.QueryRow(ctx, `
			SELECT a.id, a.tenant_id, t.slug, a.route_number, a.transfer_extension_id,
			       te.extension_number, av.id, av.provider_connection_id, av.provider, av.model,
			       COALESCE(av.voice, ''), COALESCE(av.language, 'en'),
			       COALESCE(av.system_instructions, ''), COALESCE(av.opening_message, ''),
			       av.allowed_tools, COALESCE(av.max_duration_seconds, 3600)
			FROM ai_agents a
			JOIN tenants t ON t.id = a.tenant_id
			JOIN ai_agent_versions av ON av.id = a.active_version_id
			JOIN extensions te ON te.id = a.transfer_extension_id
			WHERE t.slug = $1 AND a.route_number = $2
			  AND a.is_active = true AND a.status = 'active'
			  AND t.status = 'active'
		`, tenantSlug, routeNumber).Scan(
			&info.ID, &info.TenantID, &info.TenantSlug, &info.RouteNumber, &info.TransferExtensionID,
			&info.TransferNumber, &info.ActiveVersionID, &info.ProviderConnectionID, &info.Provider,
			&info.Model, &info.Voice, &info.Language, &info.SystemInstructions, &info.OpeningMessage,
			&toolsJSON, &info.MaxDurationSeconds,
		)
	})
	if err != nil {
		return nil, err
	}
	_ = json.Unmarshal(toolsJSON, &info.AllowedTools)
	return &info, nil
}

func (r *Repository) CreateAiSession(ctx context.Context, rec AiSessionRecord) error {
	return r.withBypass(ctx, func(tx pgx.Tx) error {
		_, err := tx.Exec(ctx, `
			INSERT INTO ai_sessions (
				id, tenant_id, call_id, agent_id, agent_version_id, provider_connection_id,
				provider_type, status, state, correlation_id, diagnostics, timing, started_at
			) VALUES ($1,$2,$3,$4,$5,$6,$7,'connecting','CREATED',$8,'{}','{}',NOW())
		`, rec.ID, rec.TenantID, rec.CallID, rec.AgentID, rec.AgentVersionID, rec.ProviderConnectionID,
			rec.ProviderType, rec.CorrelationID)
		return err
	})
}

func (r *Repository) UpdateAiSessionState(ctx context.Context, sessionID uuid.UUID, state, status string, fields map[string]any) error {
	return r.withBypass(ctx, func(tx pgx.Tx) error {
		query := `UPDATE ai_sessions SET state = $2, status = $3`
		args := []any{sessionID, state, status}
		i := 4
		if v, ok := fields["provider_session_id"]; ok {
			query += `, provider_session_id = $` + fmt.Sprint(i)
			args = append(args, v)
			i++
		}
		if v, ok := fields["diagnostics"]; ok {
			body, _ := json.Marshal(v)
			query += `, diagnostics = diagnostics || $` + fmt.Sprint(i) + `::jsonb`
			args = append(args, body)
			i++
		}
		if v, ok := fields["timing"]; ok {
			body, _ := json.Marshal(v)
			query += `, timing = timing || $` + fmt.Sprint(i) + `::jsonb`
			args = append(args, body)
			i++
		}
		if v, ok := fields["ended_at"]; ok {
			query += `, ended_at = $` + fmt.Sprint(i)
			args = append(args, v)
		}
		query += ` WHERE id = $1`
		_, err := tx.Exec(ctx, query, args...)
		return err
	})
}

func (r *Repository) GetAiSessionByCallID(ctx context.Context, callID uuid.UUID) (uuid.UUID, string, json.RawMessage, json.RawMessage, error) {
	var sessionID uuid.UUID
	var state string
	var diagnostics []byte
	var timing []byte
	err := r.withBypass(ctx, func(tx pgx.Tx) error {
		return tx.QueryRow(ctx, `
			SELECT id, state, diagnostics, timing
			FROM ai_sessions WHERE call_id = $1 ORDER BY started_at DESC LIMIT 1
		`, callID).Scan(&sessionID, &state, &diagnostics, &timing)
	})
	return sessionID, state, diagnostics, timing, err
}

func (r *Repository) GetLatestAiSessionForTenant(ctx context.Context, tenantID uuid.UUID) (uuid.UUID, uuid.UUID, string, json.RawMessage, error) {
	var sessionID uuid.UUID
	var callID uuid.UUID
	var state string
	var diagnostics []byte
	err := r.withBypass(ctx, func(tx pgx.Tx) error {
		return tx.QueryRow(ctx, `
			SELECT id, call_id, state, diagnostics
			FROM ai_sessions WHERE tenant_id = $1 ORDER BY started_at DESC LIMIT 1
		`, tenantID).Scan(&sessionID, &callID, &state, &diagnostics)
	})
	return sessionID, callID, state, diagnostics, err
}

func (r *Repository) GetProviderCredentialsEncrypted(ctx context.Context, tenantID, connID uuid.UUID) (string, string, error) {
	var enc, providerType string
	err := r.withBypass(ctx, func(tx pgx.Tx) error {
		return tx.QueryRow(ctx, `
			SELECT credentials_encrypted, provider_type
			FROM ai_provider_connections
			WHERE id = $1 AND tenant_id = $2 AND is_active = true
		`, connID, tenantID).Scan(&enc, &providerType)
	})
	return enc, providerType, err
}

func (r *Repository) WriteAiUsage(ctx context.Context, tenantID, sessionID, callID, correlationID uuid.UUID, meter, unit, source string, qty float64, idempotencyKey, providerEventID string) (bool, error) {
	inserted := false
	err := r.withBypass(ctx, func(tx pgx.Tx) error {
		tag, err := tx.Exec(ctx, `
			INSERT INTO ai_usage (
				tenant_id, session_id, call_id, provider, meter_name, quantity, unit,
				measurement_source, idempotency_key, provider_event_id, correlation_id, recorded_at
			) VALUES ($1,$2,$3,'ai',$4,$5,$6,$7,$8,$9,$10,NOW())
			ON CONFLICT (idempotency_key) WHERE idempotency_key IS NOT NULL DO NOTHING
		`, tenantID, sessionID, callID, meter, qty, unit, source, idempotencyKey, providerEventID, correlationID)
		if err != nil {
			return err
		}
		inserted = tag.RowsAffected() > 0
		return nil
	})
	return inserted, err
}

func (r *Repository) EndAiSession(ctx context.Context, sessionID uuid.UUID, state, failureCategory string) error {
	return r.UpdateAiSessionState(ctx, sessionID, state, "completed", map[string]any{
		"ended_at": time.Now().UTC(),
	})
}
